### ISSUE TITLE
Filtering predictions with conf_thres in  non_max_suppression

### ISSUE THREAD

**Issue by dheCEL** [ultralytics/ultralytics#18201]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

Originally I wanted to extract the class probabilities for a YOLOv8 model for all classes and not just the top one. While going through the code I came across the `non_max_suppression` function in `ultralytics/utils/ops.py`. 
The first time the detections are filtered based on the `conf_thres` parameter happens in [L229](https://github.com/ultralytics/ultralytics/blob/1cfe60e1d3b080943110b6d052e61ac9b35ea54b/ultralytics/utils/ops.py#L229) when filtering for candidate bounding boxes.

```python
bs = prediction.shape[0]  # batch size (BCN, i.e. 1,84,6300)
nc = nc or (prediction.shape[1] - 4)  # number of classes
nm = prediction.shape[1] - nc - 4  # number of masks
mi = 4 + nc  # mask start index
xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates
```

Then again in [L266](https://github.com/ultralytics/ultralytics/blob/1cfe60e1d3b080943110b6d052e61ac9b35ea54b/ultralytics/utils/ops.py#L266) and [L270](https://github.com/ultralytics/ultralytics/blob/1cfe60e1d3b080943110b6d052e61ac9b35ea54b/ultralytics/utils/ops.py#L270). The reason for this was already explained in issue https://github.com/ultralytics/ultralytics/issues/4075, so I'm not questioning this.

```python
if multi_label:
    i, j = torch.where(cls > conf_thres)
    x = torch.cat((box[i], x[i, 4 + j, None], j[:, None].float(), mask[i]), 1)
else:  # best class only
    conf, j = cls.max(1, keepdim=True)
    x = torch.cat((box, conf, j.float(), mask), 1)[conf.view(-1) > conf_thres]
```

What I don't understand however is the fact that the filtering with the `conf_thres` seemingly happens on logit tensors where softmax wasn't applied and therefore the filtering based on `conf_thres` doesn't make any sense to me.

In [L263](https://github.com/ultralytics/ultralytics/blob/1cfe60e1d3b080943110b6d052e61ac9b35ea54b/ultralytics/utils/ops.py#L263) of `ultralytics/utils/ops.py` we extract the `box`, `cls` and `mask` for every item in the prediction tensor. As one would expect, they have the size of `(n_detections x 4)` for `box`, `(n_detections x n_classes)` for `cls` and `(n_detections x 32)` for `mask`. 

```python
# Detections matrix nx6 (xyxy, conf, cls)
box, cls, mask = x.split((4, nc, nm), 1)
```

I expected that each element in the `cls` tensor would sum up to 1, as these should be the probabilities over all classes for a single detection, but they some up to some arbitrary number.

Do I have a false understanding here, or shouldn't this `cls` tensor be interpreted as probabilities here?

**Comment by Y-T-G** [comment]
The probabilities don't use softmax in object detection. They use sigmoid, so they don't sum up to one.

https://github.com/ultralytics/ultralytics/blob/1cfe60e1d3b080943110b6d052e61ac9b35ea54b/ultralytics/nn/modules/head.py#L131

**Comment by dheCEL** [comment]
Thank you for your answer, this clarifies a lot of things. I haven't found this activation. Makes much more sense now to me.

**Comment by glenn-jocher** [comment]
@dheCEL you're very welcome! I'm glad it makes more sense now. If you have any further questions about the code or YOLO models, feel free to ask. The community and the Ultralytics team are here to help.
